Use newline method to separate choices in form builder choices field

amajai opened the pull request:
Fix to #3001 

- Implemented the newline approach for separating choices within the form builder choices textarea as an alternative to using commas.
- If the newline isn't used in the textarea, it falls back to using commas for separating choices. 
- Made it so that it doesn't affect existing implementations that use commas.

Changelog
```
 * Add newline separation method as an alternative to using commas for separating choices in form builder choices field (Abdulmajeed Isa)
```

lb- commented:
Looking good so far, we need to now solve how users will actually enter new lines into the choices field in the page editor. 

Choices field is not a text area
-     choices = models.TextField(
- https://github.com/wagtail/wagtail/blob/5a4d929c175443a6695045ddc90fec78a93993f9/wagtail/contrib/forms/models.py#L103
- which means, by default, the user cannot enter any new lines
- we may need to provide either a default of the field using the textarea widget or set it by default - FieldPanel("choices", classname="formbuilder-choices"),
- I personally think the default should be a simple text field and we should document how to change this. 
- https://docs.wagtail.org/en/stable/reference/contrib/forms/index.html
- probably documenting a different set of panels is better for the documentation 
- please manually validate your documentation verbatim

amajai commented:
@lb- The choices field default form widget is already Textarea. I had no problem entering newlines.  Am I  doing something wrong?

lb- commented:
My bad. Thanks for checking. I'll try to run this locally and check.

lb- commented on wagtail/contrib/forms/forms.py:
```diff
@@ -23,6 +23,20 @@ class FormBuilder:
     def __init__(self, fields):
         self.fields = fields
 
+    def format_choices(self, field):
```
```suggestion
    def format_choices(self, field):
        """
        Returns a list of choices (string).
        Split the provided choices into a list, separated by new lines.
        If no new lines in the provided choices value, split by commas.
        """

```

lb- commented on wagtail/contrib/forms/forms.py:
```diff
@@ -47,25 +61,19 @@ def create_number_field(self, field, options):
     def create_checkboxes_field(self, field, options):
-        options["choices"] = [(x.strip(), x.strip()) for x in field.choices.split(",")]
+        options["choices"] = self.format_choices(field)
         options["initial"] = [x.strip() for x in field.default_value.split(",")]
```
We will likely need to update this initial value to also use the same function so that it handles new lines.

amajai replied:
@lb- May need to use a new function since the initial value requires list of values rather than list of key and value pair.

lb- replied:
could use the same function and then just get the values from the dict result maybe?

amajai replied:
Okay will try that.

lb- reviewed (changes requested):
Hey, looking great, I did a test and it works really well.
I have tested exported reports and the commas are escaped correctly.

### Next steps
* Update the help text for the choices field, current is `"Comma separated list of choices. Only applicable in checkboxes, radio and dropdown."` and likely should be changed to `"Comma or new line separated list of choices. Only applicable in checkboxes, radio and dropdown."`
* Update the `default_value` to a `models.TextField` so it can handle new lines also, will need a unit test for this case.
* Update the help text for default value, similar to above, current is `"Default value. Comma separated values supported for checkboxes."`
* Add a docstring to the `format_choices` method (i put a suggestion up)
* Ensure that the 'initial' value on checkboxes also can handle commas if new lines exist (once field is updated).
* Changing the help text and `default_value` field type will trigger a migration for the core wagtail models so let me know if you need a hand with that.

### Validation screenshots

<img width="590" alt="Screen Shot 2022-03-22 at 5 16 02 pm" src="https://user-images.githubusercontent.com/1396140/159428435-59b16f3d-2420-403d-a331-45598ab064ac.png">

<img width="1236" alt="Screen Shot 2022-03-22 at 5 16 31 pm" src="https://user-images.githubusercontent.com/1396140/159428424-fa191d90-7aa9-4779-b579-83c5ff1afb50.png">

amajai commented:
@lb- Thanks alot for the review. Will make the changes right away. Regarding triggering the core wagtail migration, does it have a different approach then the usually way, which is excuting `./manage.py makemigrations` then `./manage.py migrate`?

lb- commented:
> @lb- Thanks alot for the review. Will make the changes right away. Regarding triggering the core wagtail migration, does it have a different approach then the usually way, which is excuting `./manage.py makemigrations` then `./manage.py migrate`?

Hey, I realised you should not need to run core migrations but probably just migrations for the test app, see here for instructions.
https://docs.wagtail.org/en/stable/contributing/developing.html#running-migrations-for-the-test-app-models

This is because `AbstractFormField` is only an abstract model in Wagtail code and is used to generate a real model only in code that uses it (Wagtail installations), so the migration will happen in other's app.

amajai commented:
@lb- I think there may be a typo in the command provided in docs on running migrations for the test app models.
```
$ django-admin.py makemigrations --settings=wagtail.tests.settings
```
When I ran this command shown in the docs, I get an error stating `ModuleNotFoundError: No module named 'wagtail.tests.settings'`

After running the command with correction (change 'tests' to 'test), it work for me.

gasman commented:
@amajai Good catch! The test module was changed from `wagtail.tests` to `wagtail.test` very recently, so the stable version of the documentation is outdated now - you can see the updated instructions on the 'latest' docs at https://docs.wagtail.org/en/latest/contributing/developing.html#running-migrations-for-the-test-app-models. (I notice a couple of remaining references to `wagtail.tests` elsewhere - I'll fix those.)

lb- commented:
Will need to validate potential backwards compatibility issue. 

I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label. 

If so, this may no longer work with this change. We may need to think about providing a solution for this such as a class attribute for 'split_token' or Documen how to customise the get choices method.

amajai commented:
@lb- I have tested the backward compatibility for both TextFields (Choices and Default value form fields) and I didn't encounter any issue. I changed the Default value field back to CharField and added some values separated by commas. Then I later changed it back to TextField and make migrations. The values were still there, unaffected by the field change. The max_length parameter for the CharField and can still be the same for TextField

> I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label.

Sorry, I didn't understand what you mean.

lb- commented:
> @lb- I have tested the backward compatibility for both TextFields (Choices and Default value form fields) and I didn't encounter any issue. I changed the Default value field back to CharField and added some values separated by commas. Then I later changed it back to TextField and make migrations. The values were still there, unaffected by the field change. The max_length parameter for the CharField and can still be the same for TextField
> 
> > I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label.
> 
> Sorry, I didn't understand what you mean.



> @lb- I have tested the backward compatibility for both TextFields (Choices and Default value form fields) and I didn't encounter any issue. I changed the Default value field back to CharField and added some values separated by commas. Then I later changed it back to TextField and make migrations. The values were still there, unaffected by the field change. The max_length parameter for the CharField and can still be the same for TextField
> 
> > I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label.
> 
> Sorry, I didn't understand what you mean.

Sorry for the delayed reply, this is a non-issue, if someone is using 'checkboxes' we can assume they want more than one thing, otherwise they should use 'checkbox'

lb- commented:
> @lb- I have tested the backward compatibility for both TextFields (Choices and Default value form fields) and I didn't encounter any issue. I changed the Default value field back to CharField and added some values separated by commas. Then I later changed it back to TextField and make migrations. The values were still there, unaffected by the field change. The max_length parameter for the CharField and can still be the same for TextField
> 
> > I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label.
> 
> Sorry, I didn't understand what you mean.



> @lb- I have tested the backward compatibility for both TextFields (Choices and Default value form fields) and I didn't encounter any issue. I changed the Default value field back to CharField and added some values separated by commas. Then I later changed it back to TextField and make migrations. The values were still there, unaffected by the field change. The max_length parameter for the CharField and can still be the same for TextField
> 
> > I'd choices are saved as multi-line values, did that ever work? E.g. checkbox label or select label.
> 
> Sorry, I didn't understand what you mean.

Sorry for the delayed reply, this is a non-issue, if someone is using 'checkboxes' we can assume they want more than one thing, otherwise they should use 'checkbox'

lb- reviewed (approved):
@amajai did a bunch of tests and works great, including multiple default values (checkboxes/multi-select).

validated a few assumptions and also validated the previous (comma) behaviour).

I have rebased/sqaushed and pushed to your branch, will check the CI finishes and merge soon.

Note: I split `format_choices_or_values` into two methods, as putting that in one felt a bit awkward.
